File: app/src/cs/wordReceipt.cs
using System;
using System.Globalization;
using System.IO;
using Newtonsoft.Json.Linq;
using System.Windows.Forms;

class Program
{
    [STAThread] // Required for SaveFileDialog
    static void Main()
    {
        try
        {
            // === Load JSON form state ===
            string baseDir = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, @"..\..\..\..");
            string jsonPath = Path.Combine(baseDir, "app", "data.json");
            string jsonText = File.ReadAllText(jsonPath);
            JObject json = JObject.Parse(jsonText);

            // === Extract form values ===
            string clientName = (string)json["form"]["clientName"];
            string clientLanguage = (string)json["form"]["clientLanguage"];
            string paymentMethod = (string)json["form"]["paymentMethod"];
            double depositAmount = (double)json["form"]["depositAmount"];
            string receiptReason = (string)json["form"]["receiptReason"];
            string lawyerName = (string)json["lawyer"]["name"];

            if (string.IsNullOrEmpty(clientName) || string.IsNullOrEmpty(lawyerName))
                throw new Exception("Missing required input.");

            string lang = clientLanguage == "Français" ? "fr" : "en";
            string locale = lang == "fr" ? "fr-CA" : "en-US";
            string templatePath = Path.Combine(baseDir, "app", "templates", lang, "Receipt.docx");

            if (!File.Exists(templatePath))
                throw new FileNotFoundException("Receipt template not found.", templatePath);

            // === Create temp copy of template ===
            string tempDocPath = Path.Combine(Path.GetTempPath(), "Receipt_" + Guid.NewGuid() + ".docx");
            File.Copy(templatePath, tempDocPath, true);

            // === Prepare replacement values ===
            string formattedDate = DateTime.Today.ToString("D", CultureInfo.CreateSpecificCulture(locale));
            string formattedAmount = depositAmount.ToString("F2");

            // === Open and fill Word document ===
            Type wordType = Type.GetTypeFromProgID("Word.Application");
            dynamic wordApp = Activator.CreateInstance(wordType);
            dynamic doc = wordApp.Documents.Open(tempDocPath);

            // === Prepare reason text ===
            string reasonText;
            if (receiptReason == "consultation")
                reasonText = lang == "fr" ? "une consultation juridique" : "a legal consultation";
            else if (receiptReason == "trust")
                reasonText = lang == "fr" ? "un paiement en fidéicommis" : "a trust payment";
            else
                reasonText = receiptReason; // Use the raw value for custom/other reasons
            
            // === Replace placeholders ===
            Util.WordReplaceText(doc, "{user}", "Michel Assi-Bodje");
            Util.WordReplaceText(doc, "{reason}", reasonText);
            Util.WordReplaceText(doc, "{clientName}", clientName);
            Util.WordReplaceText(doc, "{paymentMethod}", paymentMethod);
            Util.WordReplaceText(doc, "{depositAmount}", formattedAmount);
            Util.WordReplaceText(doc, "{lawyerName}", lawyerName);
            Util.WordReplaceText(doc, "{date}", formattedDate);

            wordApp.Visible = true;
            WindowFocus.ShowWithFocus(doc.ActiveWindow);

            // Show the print dialog for the user to choose settings and print
            wordApp.Dialogs[97].Show(); // 97 = wdDialogFilePrint (Ctrl+P dialog)

            // === Ask user to export PDF ===
            using (var dialog = new SaveFileDialog())
            {
                dialog.Filter = "PDF files (*.pdf)|*.pdf";
                dialog.Title = "Save Receipt as PDF";
                dialog.FileName = DateTime.Now.ToString("yyyy-MM-dd") + "_" + clientName.Replace(" ", "-") + "_" + ".pdf";

                string projectPath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, @"..\..\..");

                string tempFilePath = Path.Combine(projectPath, "app", "latest_receipt_path.txt");
    
                if (dialog.ShowDialog() == DialogResult.OK)
                {
                    string pdfPath = dialog.FileName;
                    doc.ExportAsFixedFormat(pdfPath, 17); // 17 = wdExportFormatPDF

                    // Save the PDF path to temp file
                    File.WriteAllText(tempFilePath, pdfPath);

                    Console.WriteLine("PDF saved to: " + pdfPath);
                    Console.WriteLine("PDF path saved to: " + tempFilePath);
                    
                    // === Cleanup ===
                    doc.Close(false);
                    wordApp.Quit();
                    File.Delete(tempDocPath);
    
                    Console.WriteLine("Receipt successfully generated.");
                }
                else
                {
                    Util.CallWordCleaner(tempDocPath, tempFilePath);
                }
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine("Error: " + ex.Message);
        }
    }
}

File: app/src/wordReceipt.py
from office_utils import *
from parse_json import *
import shutil
import tempfile
import pythoncom
from datetime import datetime
from tkinter import Tk, filedialog

def process_word_receipt(data):
    try:
        form, _, lawyer = split_data(data)

        client_name = form.get("clientName", "")
        client_language = get_language(data)
        payment_method = form.get("paymentMethod", "")
        deposit_amount = float(form.get("depositAmount", 0))
        receipt_reason = form.get("receiptReason", "")
        lawyer_id = lawyer.get("id", "")
        lawyer_name = lawyer.get("name", "")

        lang = "fr" if client_language.startswith("Français") else "en"
        template_path = os.path.join(TEMPLATES_DIR, lang, "Receipt.docx")
        if not os.path.exists(template_path):
            raise FileNotFoundError(f"Receipt template not found: {template_path}")

        # Create temp file in a reliable location - FIXED
        temp_dir = tempfile.gettempdir()
        temp_doc_path = os.path.join(temp_dir, f"Receipt_{os.urandom(4).hex()}.docx")
        
        # Ensure temp directory exists
        os.makedirs(temp_dir, exist_ok=True)
        shutil.copy(template_path, temp_doc_path)
        
        # Verify the temp file was created
        if not os.path.exists(temp_doc_path):
            raise FileNotFoundError(f"Failed to create temporary file: {temp_doc_path}")

        # Formatting
        formatted_date = format_date(datetime.today(), lang)
        formatted_amount = f"{deposit_amount:.2f}"
        lawyer_string = get_lawyer_string(lawyer_name, lawyer_id)

        # Prepare reason text
        reason_text = {
            "consultation": "une consultation juridique" if lang == "fr" else "a legal consultation",
            "trust": "un paiement en fidéicommis" if lang == "fr" else "a trust payment"
        }.get(receipt_reason, receipt_reason)

        replacements = {
            "{user}": "Michel Assi-Bodje",
            "{reason}": reason_text,
            "{clientName}": client_name,
            "{paymentMethod}": payment_method,
            "{depositAmount}": formatted_amount,
            "{lawyerName}": lawyer_string,
            "{date}": formatted_date
        }
        
        # Init Word COM with better error handling
        pythoncom.CoInitialize()
        try:
            word = COM.DispatchEx("Word.Application")
            word.Visible = False  # Keep hidden initially
            doc = word.Documents.Open(temp_doc_path)
        except Exception as word_error:
            raise Exception(f"Failed to open Word document: {str(word_error)}")

        # Process replacements
        try:
            for placeholder, replacement in replacements.items():
                word_replace_text(doc, placeholder, replacement)
        except Exception as replace_error:
            raise Exception(f"Failed to process document replacements: {str(replace_error)}")

        # Show Word window after processing
        word.Visible = True
        focus_office_window(doc.ActiveWindow)

        # Print dialog (File > Print)
        try:
            word.Dialogs(88).Show()  # 88 = wdDialogFilePrint
        except Exception as print_error:
            print(f"Warning: Print dialog failed: {print_error}", file=sys.stderr)

        # Save as PDF dialog with better error handling
        try:
            root = Tk()
            root.withdraw()
            root.lift()  # Bring to front
            root.attributes('-topmost', True)  # Keep on top
            
            default_filename = f"{datetime.today():%Y-%m-%d}_{client_name.replace(' ', '-')}.pdf"
            
            pdf_path = filedialog.asksaveasfilename(
                defaultextension=".pdf",
                filetypes=[("PDF files", "*.pdf")],
                title="Save Receipt as PDF",
                initialfile=default_filename
            )
            root.destroy()
        except Exception as dialog_error:
            raise Exception(f"Failed to show save dialog: {str(dialog_error)}")

        if pdf_path:
            try:
                # Validate and sanitize the PDF path (similar to contract)
                pdf_path = os.path.normpath(pdf_path)
                pdf_dir = os.path.dirname(pdf_path)
                
                # Check if directory exists, create if it doesn't
                if not os.path.exists(pdf_dir):
                    os.makedirs(pdf_dir, exist_ok=True)
                
                # Check if we can write to the directory
                if not os.access(pdf_dir, os.W_OK):
                    raise Exception(f"No write permission to directory: {pdf_dir}")
                
                # Sanitize filename
                filename = os.path.basename(pdf_path)
                invalid_chars = '<>:"/\\|?*'
                for char in invalid_chars:
                    if char in filename:
                        filename = filename.replace(char, '_')
                
                pdf_path = os.path.join(pdf_dir, filename)
                
                print(f"Attempting to export PDF to: {pdf_path}", file=sys.stderr)
                
                doc.ExportAsFixedFormat(pdf_path, 17)  # 17 = wdExportFormatPDF
                              
            except Exception as export_error:
                raise Exception(f"PDF export failed: {str(export_error)}")
            finally:
                # Always clean up Word
                try:
                    doc.Close(False)
                    word.Quit()
                
                    return {
                        "status": "success",
                        "message": "Word receipt successfully created.",
                        "pdf_path": pdf_path
                    }
                except:
                    pass
                # Clean up temp file
                try:
                    if os.path.exists(temp_doc_path):
                        os.remove(temp_doc_path)
                except:
                    # If immediate cleanup fails, use the cleaner
                    call_cleaner_async(temp_doc_path)
        else:
            # User cancelled - clean up
            try:
                doc.Close(False)
                word.Quit()
            except:
                pass
            call_cleaner_async(temp_doc_path)

            return {
                "status": "cancelled",
                "message": "User cancelled receipt export to PDF."
            }
            
    except Exception as e:
        alert_error(f"Error: {e}")
        return {
            "status": "error",
            "error": str(e),
            "message": f"Failed to create receipt: {str(e)}"
        }
    finally:
        try:
            pythoncom.CoUninitialize()
        except:
            pass

# Backward compatibility
def main():
    try:
        data = read_json()
        result = process_word_receipt(data)
        if result.get("error"):
            print(f"Error: {result['error']}")
        else:
            print(result.get("message", "Word receipt processed"))
    except Exception as e:
        print(f"Error: {str(e)}")

if __name__ == "__main__":
    main()